src/features/toast/ToastRenderer.tsx

```diff
@@ -0,0 +1,80 @@
+import React, { useRef } from 'react'
+import { useSelector } from 'react-redux'
+import { ToastContainer, toast } from 'react-toastify'
+
+import { Run, RunStatus } from '../../qri/run'
+import Toast from './Toast'
+import { selectRunsFromEventLog } from '../workflow/state/workflowState'
+
+
+import 'react-toastify/dist/ReactToastify.css'
+
+const ToastRenderer: React.FC<{}> = () => {
+
+  const runs = useSelector(selectRunsFromEventLog)
+
+  // use a ref to store the workflow status so we can
+  // determine when it changes and update an already visible toast
+  const runsRef = useRef({} as Record<string, RunStatus>)
+
+  // artifically delay the update to 'success', ensures that
+  // user will see 'running' for at least 3 seconds even if the
+  // workflow runs quickly
+  const SUCCESS_START_DELAY = 3000
+
+  // after 'success', close the toast 4 seconds later
+  const SUCCESS_CLOSE_DELAY = 4000
+
+  // gets all of the known runs from state
+  // iterate over each one and show/update/dismiss a toast
+  runs.forEach((run: Run) => {
+    const { id, status } = run
+    // if deploying, make sure there is not already a toast showing for this workflowId,
+    // then create a new toast
+
+    switch(status) {
+      case 'running':
+        // only create a new toast if there's no record for this id in runsRef
+        if(!runsRef.current[id]) {
+          runsRef.current[id] = status
+          toast(<Toast message='Running Workflow...' type='running' id={id} />, {
+            toastId: id,
+            autoClose: false,
+          })
+        }
+      break
+      case 'succeeded':
+        // if deploy is successful, update the toast, and autoClose
+        if (runsRef.current[id] === 'running') {
+          runsRef.current[id] = status
+          setTimeout(() => {
+            toast.update(id, {
+              render: (
+                <Toast message='Success!' type='succeeded' id={id}  />
+              ),
+            })
+            setTimeout(() => {
+              delete runsRef.current[id]
+              toast.dismiss(id)
+            }, SUCCESS_START_DELAY)
+          }, SUCCESS_CLOSE_DELAY)
+        }
```

Comment:
rather than having a `setTimeout` to dismiss the toast, shouldn't we instead pass in a `autoClose` value of `SUCCESS_CLOSE_DELAY` and an an `onClose` callback that deletes the id from the `runsRef` object?

```
if (runsRef.current[id] === 'running') {
    runsRef.current[id] = status
    setTimeout(() => {
        toast.update(id, {
            render: <Toast message='Success!' type='succeeded' id={id}  />,
            autoClose: SUCCESS_CLOSE_DELAY,
            onClose: () => { delete runsRef.current[id] }
        })}, SUCCESS_START_DELAY)
}
```

The `setTimeout` to delay the start I can't think of a less complicated solution for the moment.


Regardless,`SUCCESS_START_DELAY` is supposed to be the time before we show the "Success", and `SUCCESS_CLOSE_DELAY` is for the time we see the "Success" toast before it closes, correct? If so, these are backwards in the code.

Reply:
Tried this method out, it works great and is much easier to read.